File: pyversion/32. longest parentheses.py
def longestValidParentheses(s):
    n = len(s)
    longest = 0
    st = []
    for i in range(n):
        if s[i] == '(':
            st.append(i)
        else:
            if (len(st) != 0):
                if s[st[-1]] == '(':
                    st.pop()
                else:
                    st.append(i)
            else:
                st.append(i)

    if len(st) == 0:
        longest = n
    else:
        a = n
        b = 0
        while len(st) != 0:
            b = st[-1]
            st.pop()
            longest = max(longest, a - b - 1)
            a = b
        longest = max(longest, a)
    return longest


if __name__ == '__main__':
    print(longestValidParentheses('()(((())'))

File: cppversion/32. longest parentheses.cpp
#include <iostream>
#include <stack>
#include <string>

int longestValidParentheses(std::string s)
{
    int n = s.length(), longest = 0;
    std::stack<int> st;
    for (int i = 0; i < n; i++)
    {
        if (s[i] == '(')
            st.push(i);
        else
        {
            if (!st.empty())
            {
                if (s[st.top()] == '(')
                    st.pop();
                else
                    st.push(i);
            }
            else
                st.push(i);
        }
    }
    if (st.empty())
        longest = n;
    else
    {
        int a = n, b = 0;
        while (!st.empty())
        {
            b = st.top();
            st.pop();
            longest = std::max(longest, a - b - 1);
            a = b;
        }
        longest = std::max(longest, a);
    }
    return longest;
}

int main()
{
    std::cout << longestValidParentheses("()(((())") << std::endl;
    std::cin.get();
}